sdk/dotnet/State/ArtifactoryRemoteStateReferenceArgs.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+using Pulumi.Serialization;
+
+namespace Pulumi.Terraform.State
+{
+    /// <summary>
+    /// The configuration options for a Terraform Remote State stored in the Artifactory backend. 
+    /// </summary>
+    public class ArtifactoryRemoteStateReferenceArgs : RemoteStateReferenceArgs
```

Comment:
@pgavlin See https://github.com/pulumi/pulumi-terraform/pull/517#discussion_r368624262

Reply:
This definitely is not something that the schema supports. Do we need to use an approach like this?

Reply:
How can we model sum types? How is the [TypeScript version](https://github.com/pulumi/pulumi-terraform/blob/master/sdk/nodejs/state/remoteStateReference.ts#L33) represented in the schema?

Reply:
We can't at the moment. FWIW, this is the only SDK that uses this pattern.

Reply:
Should we remodel as N resources with N independent args types?

Reply:
That would work. We could also use one resource with one arg type that supports all of the properties needed for the various backends, then use overlays for a more ergonomic experience.

Reply:
I think separate resources make for a better user experience. I changed the current implementation to the N resources model.

Reply:
We should also change the type tokens s.t. they reflect the resource type. Note that this will require changing the provider implementation accordingly.

Reply:
That sounds like a breaking change and 2.0 then?
Do they have to match?
Should we proceed with the original implementation for now and leave this for when the resource model is there, if it's breaking anyway?

Reply:
Besides, we'd need to fix master vs dotnet branches before that...

Reply:
We don't have to make that a breaking change: we can continue to accept the original type token as well as the new, more specific tokens.

We could also treat the union args as overlays and let the schema describe the union of the args type. If that's not clear, lmk and I'll put together an example.

Reply:
Overlays mean hand-written specific types that are mapped to one internal union type?

Reply:
That's right.

Reply:
Wouldn't that be exactly equivalent to my current PR in terms of public API and resource type token?

Reply:
It'd be very close, yeah. There would be one additional resource + args type for the single actual resource (i.e. `RemoteStateReference`).

Reply:
Would `RemoteStateReference` be the base class then (but not the args)?

Reply:
It could be. I don't think it has to be, but if that makes the most sense for the SDK then that's probably what we should do.

Reply:
I gave it a thought and it feels very awkward to me. A base class `RemoteStateReference` class that nobody should actually use (they should use overlays instead) and monster `RemoteStateReferenceArgs` which includes all properties from all args like `IamEndpoint` and `ResourceGroupName`...
@CyrusNajmabadi could you join the discussion above and see what you think?